addons/mail/static/tests/owl/component/chat_window_manager_tests.js

```diff
@@ -601,6 +601,423 @@ QUnit.test('chat window: open / close', async function (assert) {
     assert.verifySteps(['rpc:channel_minimize']);
 });
 
+QUnit.test('chat window: state conservation on toggle home menu', async function (assert) {
+    assert.expect(10);
+
+    // HELPERS SETUP & START
+
+    Object.assign(this.data.initMessaging, {
+        channel_slots: {
+            channel_channel: [{
+                channel_type: 'channel',
+                id: 20,
+                is_minimized: false,
+                name: "General",
+                state: 'open',
+                uuid: 'channel-20-uuid',
+            }],
+        },
+    });
+    const channelId = 20;
+    const localChannelId = `mail.channel_${channelId}`;
+    await this.start({
+        async mockRPC(route, args) {
+            if (args.method === 'channel_fetch_preview') {
+                return [{
+                    id: 20,
+                    last_message: {
+                        author_id: [7, "Demo"],
+                        body: "<p>test</p>",
+                        channel_ids: [20],
+                        id: 100,
+                        message_type: 'comment',
+                        model: 'mail.channel',
+                        res_id: 20,
+                    },
+                }];
+            } else if (args.method === 'channel_fold') {
+                return;
+            } else if (args.method === 'message_fetch') {
+                return [...Array(20).keys()].map(i => {
+                    return {
+                        author_id: [11, "Demo"],
+                        body: "<p>body</p>",
+                        channel_ids: [20],
+                        date: "2019-04-20 10:00:00",
+                        id: i + 10,
+                        message_type: 'comment',
+                        model: 'mail.channel',
+                        record_name: 'General',
+                        res_id: i + 1,
+                    }
+                });
+            }
+            return this._super(...arguments);
+        },
+    });
+
+    // Open thread
+    document
+        .querySelector(`.o_MessagingMenu_toggler`)
+        .click()
+    ;
+    await testUtils.nextTick(); // re-render
+    document
+        .querySelector(`
+            .o_MessagingMenu_dropdownMenu
+            .o_ThreadPreviewList_preview`)
+        .click();
+    await testUtils.nextTick(); // re-render
+
+    // SETUP INITIAL STATE
+    // - Set a scroll position to chat window
+    document.querySelector(`.o_MessageList.o_Thread_messageList`).scrollTop = 142;
+
+    // - Set html content of the composer of the chat window
+    let composerTextInput = document.querySelector(`
+        .o_ComposerTextInput
+        > .note-editor
+        > .note-editing-area
+        > .note-editable`);
+    composerTextInput.innerHTML = "XDU for the win !";
+    composerTextInput.dispatchEvent(new window.KeyboardEvent('input'));
+
+    // - Set attachments of the composer
+    const composerLocalId = Object.keys(self.store.state.composers)[0];
```

Comment:
Not that great to make changes to the store manually.
Would be better to make a file and mock the upload (this could be a test in itself).

To make a test, can use web/test_utils_file.js function `createFile`.
This file has some helpers for drag/drop such files, but this feature hasn't been re-implemented yet.
Could be great to see how to insert this file in `<input>`, so that it triggers `change` event and follows the usual flow of uploading a file as an attachment.

Reply:
If I understand well, you want me to add another test to check the same process but with a real attachment process instead of just manipulating the store to fake there are attachments ?

Reply:
Make a test that simply adds an attachment to the composer.
Then, this test can add the attachment in the composer the same way.

Reply:
Apparently there is a technical limitation there. You cannot set something else than a FileList object to the files input and FileList is a readonly object, you can't set anything on it and cannot create an instance of it... So I ended up directly calling the _onChangeAttachment handler method, is that ok ?

Reply:
Indeed, `FileList` is a readonly object, but maybe such a `FileList` can be created from a `File`/`Blob`, like this:
https://stackoverflow.com/questions/47119426/how-to-set-file-objects-and-length-property-at-filelist-object-where-the-files-a/47172409#47172409

If this works, then it would be very helpful to have a function like `inputFile($input, file)` in test_utils_file.js.
Test would simply have to use `createFile()` and `inputFile()` helpers.

Reply:
(in the link, consider `DataTransfer` and file is created from `createFile()`)

Reply:
But this works in discuss because I can access the component from the test. I don't know how to manage that in chat window manager tests..